bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/Tr064RootHandler.java

```diff
@@ -0,0 +1,391 @@
+        final ScheduledFuture<?> pollFuture = this.pollFuture;
+        if (pollFuture != null) {
+            pollFuture.cancel(true);
+            this.pollFuture = null;
+        }
+        final ScheduledFuture<?> phonebookFuture = this.phonebookFuture;
+        if (phonebookFuture != null) {
+            phonebookFuture.cancel(true);
+            this.phonebookFuture = null;
+        }
+    }
+
+    /**
+     * install the polling
+     */
+    private void installPolling() {
+        uninstallPolling();
+        pollFuture = scheduler.scheduleWithFixedDelay(this::poll, 0, config.refresh, TimeUnit.SECONDS);
+        if (config.phonebookInterval > 0) {
+            phonebookFuture = scheduler.scheduleWithFixedDelay(this::retrievePhonebooks, 0, config.phonebookInterval,
+                    TimeUnit.SECONDS);
+        }
+    }
+
+    private void retrievePhonebooks() {
+        String serviceId = "urn:X_AVM-DE_OnTel-com:serviceId:X_AVM-DE_OnTel1";
+        scpdUtil.getDevice("").ifPresent(deviceType -> deviceType.getServiceList().stream()
+                .filter(service -> service.getServiceId().equals(serviceId)).findFirst().ifPresent(scpdService -> {
+                    try {
+                        SOAPMessage soapMessagePhonebookList = soapConnector.doSOAPRequest(scpdService,
+                                "GetPhonebookList", Collections.emptyMap());
+                        SOAPValueConverter soapValueConverter = new SOAPValueConverter(httpClient);
+                        phonebooks = soapValueConverter
+                                .getStateFromSOAPValue(soapMessagePhonebookList, "NewPhonebookList", null)
+                                .map(phonebookList -> Arrays.stream(phonebookList.toString().split(",")).map(index -> {
+                                    try {
+                                        SOAPMessage soapMessageURL = soapConnector.doSOAPRequest(scpdService,
+                                                "GetPhonebook", Map.of("NewPhonebookID", index));
+                                        return soapValueConverter
+                                                .getStateFromSOAPValue(soapMessageURL, "NewPhonebookURL", null)
+                                                .map(url -> (Phonebook) new Tr064PhonebookImpl(httpClient,
+                                                        url.toString()))
+                                                .orElse(null);
+                                    } catch (Tr064CommunicationException e) {
+                                        logger.warn("Failed to get phonebook with index {}:", index, e);
+                                    }
+                                    return null;
+                                }).collect(Collectors.toList())).orElse(Collections.emptyList());
+                    } catch (Tr064CommunicationException e) {
+                        phonebooks = Collections.emptyList();
+                    }
+                }));
```

Comment:
I tried something similiar before. Unfortunately this will not work. `scpdService` is no longer available in l. 354 if you re-arrange the code like that.

Reply:
I refactored the code, WDYT?

Reply:
Looks better than before at least. 😄